linode/resource_linode_lke_cluster.go

```diff
@@ -406,6 +414,150 @@ func reconcileLKEClusterPoolSpecs(poolSpecs []linodeLKEClusterPoolSpec, pools []
+				}
+			}
+
+			if ready {
+				log.Printf("[DEBUG] finished waiting for LKE Cluster (%d) Pool (%d) to be ready", clusterID, poolID)
+				wg.Done()
+				return
+			}
+		}
+	}
+}
+
+func waitForClusterPoolsToStartRecycle(ctx context.Context, client *linodego.Client, clusterID int, pools []linodego.LKEClusterPool) (<-chan int, <-chan error) {
+	clusterInstances := make(map[int]int)
+	poolInstances := make(map[int]map[int]struct{}, len(pools))
+	for _, pool := range pools {
+		poolInstances[pool.ID] = make(map[int]struct{}, len(pool.Linodes))
+		for _, instance := range pool.Linodes {
+			poolInstances[pool.ID][instance.InstanceID] = struct{}{}
+			clusterInstances[instance.InstanceID] = pool.ID
+		}
+	}
+
+	poolRecyclesCh := make(chan int)
+	errCh := make(chan error)
+
+	eventTicker := time.NewTicker(200 * time.Millisecond)
+	lastEventID := 0
+
+	go func() {
+		defer eventTicker.Stop()
+		defer close(poolRecyclesCh)
+		defer close(errCh)
+
+		for len(clusterInstances) != 0 {
+			select {
+			case <-ctx.Done():
+				log.Printf("[ERROR] timed out waiting for all original nodes of LKE Cluster (%d) to be deleted (%d remaining)\n", clusterID, len(clusterInstances))
+				return
+
+			case <-eventTicker.C:
+				filterBytes, _ := json.Marshal(map[string]interface{}{
+					"+order_by": "created",
+					"+order":    "desc",
+					"+gte":      lastEventID,
+
+					"entity.type": string(linodego.EntityLinode),
+				})
+
+				events, err := client.ListEvents(ctx, linodego.NewListOptions(1, string(filterBytes)))
+				if err != nil {
+					errCh <- err
+					return
+				}
+
+				if len(events) != 0 {
+					lastEventID = events[0].ID
+				}
+
+				for _, event := range events {
+					if event.Action != linodego.ActionLinodeDelete {
+						continue
+					}
+					id := int(event.Entity.ID.(float64))
+					poolID, ok := clusterInstances[id]
+					if !ok {
+						continue
+					}
+
+					delete(clusterInstances, id)
+					delete(poolInstances[poolID], id)
+					log.Printf("[DEBUG] finished waiting for LKE Cluster (%d) Pool (%d) Node (%d) to be deleted\n", clusterID, poolID, id)
+
+					if len(poolInstances[poolID]) == 0 {
+						// all original instances for this pool have been deleted
+						delete(poolInstances, poolID)
+						log.Printf("[DEBUG] finished waiting for all nodes in LKE Cluster (%d) Pool (%d) to be recreated\n", clusterID, poolID)
+						poolRecyclesCh <- poolID
+					}
+				}
+			}
+		}
+	}()
+	return poolRecyclesCh, errCh
+}
+
+func recycleLKECluster(ctx context.Context, client *linodego.Client, id int, pools []linodego.LKEClusterPool) error {
+	if err := client.RecycleLKEClusterNodes(ctx, id); err != nil {
+		return fmt.Errorf("failed to recycle LKE Cluster (%d): %s", id, err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	poolRecyclesCh, errCh := waitForClusterPoolsToStartRecycle(ctx, client, id, pools)
+	poolsRecycled := new(sync.WaitGroup)
+	poolsRecycled.Add(len(pools))
+	poolsRecycledCh := waitGroupCh(poolsRecycled)
```

Comment:
Is this really necessary? We can discuss later.

Reply:
It does make control flow easier in my opinion, handling the wg in the same place as the error channels. I'm open to suggestions though, the whole thing is a bit complex.